### PR TITLE
feat(ginkgo) New integration with ginkgo framework

### PR DESCRIPTION
**Pull request by holyhope**
Signed-off-by: Pierre PÉRONNET <pierre.peronnet@datadoghq.com>

### CODE REVIEW

File: ginkgo/ginkgo.go [hunk 1]
```diff
@@ -0,0 +1,362 @@
...
+	initSetup()
+
+	return ginkgo.PIt(text, callback...)
+}
+
+func XIt(text string, callback ...interface{}) bool {
+	initSetup()
+
+	return ginkgo.XIt(text, callback...)
+}
```
Comment (albertvaka) [inline ginkgo/ginkgo.go:139]: Don't we want to create spans on these two cases? If it is intentional, can you add a comment explaining why?

File: ginkgo/ginkgo.go [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+
+func By(text string, callback ...func()) {
+	initSetup()
+
+	test := currentSuite.Snapshot().RegisterTest()
+
+	if len(callback) == 0 {
+		callback = []func(){func() {}}
+	}
```
Comment (albertvaka) [inline ginkgo/ginkgo.go:182]: Why is this needed?

File: ginkgo/internal/context.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+var testCountContextKey = "testCountContextKey"
+
+func WithTestCount(ctx context.Context, count int64) context.Context {
+	return context.WithValue(ctx, &testCountContextKey, &count)
+}
+
+func IncTestCount(ctx context.Context) int64 {
+	countUntyped := ctx.Value(&testCountContextKey)
+	if countUntyped == nil {
+		panic("IncTestCount: no test count")
```
Comment (albertvaka) [inline ginkgo/internal/context.go:61]:
```suggestion
		panic(fmt.Errorf("IncTestCount: no test count in %+v", ctx))
```
?

File: ginkgo/internal/context.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		panic("IncTestCount: test count is not a pointer")
+	}
+
+	return atomic.AddInt64(count, 1)
+}
+
+func DecTestCount(ctx context.Context) int64 {
+	countUntyped := ctx.Value(&testCountContextKey)
+	if countUntyped == nil {
+		panic(fmt.Errorf("IncTestCount: no test count in %+v", ctx))
```
Comment (albertvaka) [inline ginkgo/internal/context.go:75]:
```suggestion
		panic(fmt.Errorf("DecTestCount: no test count in %+v", ctx))
```

File: ginkgo/internal/context.go [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	count, ok := countUntyped.(*int64)
+	if !ok {
+		panic("IncTestCount: test count is not a pointer")
```
Comment (albertvaka) [inline ginkgo/internal/context.go:80]:
```suggestion
		panic("DecTestCount: test count is not a pointer")
```

File: ginkgo/internal/context.go [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	return atomic.AddInt64(count, -1)
+}
+
+func GetTestCount(ctx context.Context) int64 {
+	countUntyped := ctx.Value(&testCountContextKey)
+	if countUntyped == nil {
+		panic(fmt.Errorf("IncTestCount: no test count in %+v", ctx))
```
Comment (albertvaka) [inline ginkgo/internal/context.go:89]:
```suggestion
		panic(fmt.Errorf("GetTestCount: no test count in %+v", ctx))
```

File: ginkgo/internal/context.go [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	count, ok := countUntyped.(*int64)
+	if !ok {
+		panic("IncTestCount: test count is not a pointer")
```
Comment (albertvaka) [inline ginkgo/internal/context.go:94]:
```suggestion
		panic("GetTestCount: test count is not a pointer")
```

File: ginkgo/internal/suite.go [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+}
+
+func (s *Snapshot) RegisterTest() TestCase {
+	return newTest(
+		s.ctx,
+		ddtesting.WithSpanOptions(tracer.Tag("test.framework", s.FrameworkName)),
+	)
+}
+
+func (s *SuiteTest) Close() error {
```
Comment (albertvaka) [inline ginkgo/internal/suite.go:164]: Future improvement: we don't yet have suite-level visibility in dd-sdk-go-testing but this would be the place to send a `test_suite_end` event with the duration of the whole suite.

**Review by albertvaka — APPROVED** [review]
I left some questions and nits but LGTM.

File: ginkgo/ginkgo.go (hunk 1 above)
Comment (holyhope) [reply ginkgo/ginkgo.go:139]: You are right I should explain why. What about the following comment:

```golang
// no need to register this test since it will be ignored
// see pending spec: https://onsi.github.io/ginkgo/#pending-specs
```

Comment (albertvaka) [reply ginkgo/ginkgo.go]: LGTM

File: ginkgo/ginkgo.go (hunk 2 above)
Comment (albertvaka) [reply ginkgo/ginkgo.go:198]: Makes sense 👍
